File: cloud/shared/src/quota.js
// Per-tenant daily token/request quota enforcement.
//
// Counters live in Redis under
//   quota:t:<tenantId>:<YYYY-MM-DD>:requests
//   quota:t:<tenantId>:<YYYY-MM-DD>:tokens
// with a 48h TTL so a slow midnight wrap-around can never lose data.
//
// Day boundary is computed in UTC+8 (Asia/Shanghai) to match user expectations
// for a domestic-China product. Tokens are counted post-success via
// incrementTokenUsage(); requests are counted pre-attempt so failed/aborted
// requests still consume quota.

import { getRedis } from './redis.js';
import { query } from './db.js';
import { RateLimitError } from './errors.js';

const PLAN_QUOTAS = {
  free:       { tokens: 100_000,     requests: 1_000 },
  starter:    { tokens: 1_000_000,   requests: 10_000 },
  pro:        { tokens: 10_000_000,  requests: 100_000 },
  enterprise: { tokens: null,        requests: null },     // unlimited
};

const QUOTA_CFG_CACHE_TTL_SEC = 300;
const COUNTER_TTL_SEC = 48 * 3600;

function todayKeyCn() {
  // ISO date in Asia/Shanghai (UTC+8). No external dep — just shift then slice.
  const cn = new Date(Date.now() + 8 * 3600 * 1000);
  return cn.toISOString().slice(0, 10);
}

async function fetchLimits(tenantId, plan) {
  const redis = getRedis();
  const cacheKey = `quota:cfg:${tenantId}`;
  const cached = await redis.get(cacheKey);
  if (cached) return JSON.parse(cached);

  const { rows } = await query(
    `SELECT daily_token_limit, daily_request_limit FROM tenant_quotas WHERE tenant_id = $1`,
    [tenantId]
  );
  const defaults = PLAN_QUOTAS[plan] || PLAN_QUOTAS.free;
  const limits = {
    tokens:   rows[0]?.daily_token_limit   != null ? Number(rows[0].daily_token_limit)   : defaults.tokens,
    requests: rows[0]?.daily_request_limit != null ? Number(rows[0].daily_request_limit) : defaults.requests,
  };
  await redis.set(cacheKey, JSON.stringify(limits), 'EX', QUOTA_CFG_CACHE_TTL_SEC);
  return limits;
}

export async function enforceDailyQuota(ctx) {
  const limits = await fetchLimits(ctx.tenantId, ctx.plan);
  if (limits.requests == null && limits.tokens == null) return;

  const redis = getRedis();
  const date = todayKeyCn();
  const reqKey = `quota:t:${ctx.tenantId}:${date}:requests`;
  const tokKey = `quota:t:${ctx.tenantId}:${date}:tokens`;

  // Pre-check tokens: we don't know how many this request will consume,
  // so we reject if we're already at/over the limit.
  if (limits.tokens != null) {
    const used = Number(await redis.get(tokKey)) || 0;
    if (used >= limits.tokens) {
      throw new RateLimitError('Daily token quota exceeded', {
        scope: 'daily_tokens', limit: limits.tokens, used, window: 'today_cn_8',
      });
    }
  }

  if (limits.requests != null) {
    const pipe = redis.pipeline();
    pipe.incr(reqKey);
    pipe.expire(reqKey, COUNTER_TTL_SEC);
    const results = await pipe.exec();
    const count = results[0][1];
    if (count > limits.requests) {
      throw new RateLimitError('Daily request quota exceeded', {
        scope: 'daily_requests', limit: limits.requests, observed: count, window: 'today_cn_8',
      });
    }
  }
}

export async function incrementTokenUsage(tenantId, totalTokens) {
  if (!totalTokens || totalTokens <= 0) return;
  const redis = getRedis();
  const date = todayKeyCn();
  const tokKey = `quota:t:${tenantId}:${date}:tokens`;
  const pipe = redis.pipeline();
  pipe.incrby(tokKey, Math.floor(totalTokens));
  pipe.expire(tokKey, COUNTER_TTL_SEC);
  await pipe.exec();
}

export async function invalidateQuotaCache(tenantId) {
  await getRedis().del(`quota:cfg:${tenantId}`);
}

export async function readQuotaUsage(tenantId, plan) {
  const limits = await fetchLimits(tenantId, plan);
  const redis = getRedis();
  const date = todayKeyCn();
  const [reqUsed, tokUsed] = await Promise.all([
    redis.get(`quota:t:${tenantId}:${date}:requests`),
    redis.get(`quota:t:${tenantId}:${date}:tokens`),
  ]);
  return {
    date,
    requests: { used: Number(reqUsed) || 0, limit: limits.requests },
    tokens:   { used: Number(tokUsed) || 0, limit: limits.tokens },
  };
}

export { PLAN_QUOTAS };

File: internal/quota/quota.go
// Package quota implements per-tenant rate limiting and daily request/token
// caps. Counters live in Redis with 48 h TTL; the day key uses Asia/Shanghai
// so the rollover matches operator expectations.
package quota

import (
	"context"
	"encoding/json"
	"fmt"
	"strconv"
	"time"

	"github.com/amaozhao/lazirouter/internal/db"
	"github.com/amaozhao/lazirouter/internal/errs"
	"github.com/amaozhao/lazirouter/internal/redisx"
	"github.com/amaozhao/lazirouter/internal/tenantctx"
)

// PlanQuota is the default daily cap per plan. nil means unlimited.
type PlanQuota struct {
	Tokens   *int64 `json:"tokens"`
	Requests *int   `json:"requests"`
}

var planDefaults = map[string]PlanQuota{
	"free":       {ptrI64(100_000), ptrInt(1_000)},
	"starter":    {ptrI64(1_000_000), ptrInt(10_000)},
	"pro":        {ptrI64(10_000_000), ptrInt(100_000)},
	"enterprise": {nil, nil},
}

// Defaults exposes the plan map for read-only callers (admin UI).
func Defaults() map[string]PlanQuota { return planDefaults }

func todayKeyCN() string {
	t := time.Now().UTC().Add(8 * time.Hour) // CST = UTC+8
	return t.Format("2006-01-02")
}

func reqKey(tenantID int64) string {
	return fmt.Sprintf("quota:%d:%s:req", tenantID, todayKeyCN())
}
func tokKey(tenantID int64) string {
	return fmt.Sprintf("quota:%d:%s:tok", tenantID, todayKeyCN())
}
func cfgKey(tenantID int64) string {
	return fmt.Sprintf("quota:cfg:%d", tenantID)
}

// effective returns the actual cap (override or plan default) for a tenant.
type effectiveCfg struct {
	TokenLimit   *int64 `json:"daily_token_limit"`
	RequestLimit *int   `json:"daily_request_limit"`
}

func loadEffective(ctx context.Context, tenantID int64, plan string) (*effectiveCfg, error) {
	rc := redisx.C()
	if cached, err := rc.Get(ctx, cfgKey(tenantID)).Result(); err == nil {
		var c effectiveCfg
		if err := json.Unmarshal([]byte(cached), &c); err == nil {
			return &c, nil
		}
	}
	var c effectiveCfg
	err := db.Pool().QueryRow(ctx, `
		SELECT daily_token_limit, daily_request_limit
		FROM tenant_quotas WHERE tenant_id = $1
	`, tenantID).Scan(&c.TokenLimit, &c.RequestLimit)
	if err != nil {
		// no override → fall back to plan defaults
		pd := planDefaults[plan]
		c.TokenLimit = pd.Tokens
		c.RequestLimit = pd.Requests
	} else {
		// fill nulls with plan defaults
		pd := planDefaults[plan]
		if c.TokenLimit == nil {
			c.TokenLimit = pd.Tokens
		}
		if c.RequestLimit == nil {
			c.RequestLimit = pd.Requests
		}
	}
	buf, _ := json.Marshal(c)
	_ = rc.Set(ctx, cfgKey(tenantID), buf, 5*time.Minute).Err()
	return &c, nil
}

// Enforce pre-checks the request count and the (already-incremented) token
// count for today. Increments the request counter atomically. Returns
// RateLimitError when over budget.
func Enforce(ctx context.Context, t *tenantctx.TenantContext) error {
	cfg, err := loadEffective(ctx, t.TenantID, t.Plan)
	if err != nil {
		return err
	}
	rc := redisx.C()
	// 1) pre-check tokens (no increment — happens after upstream returns)
	if cfg.TokenLimit != nil {
		raw, _ := rc.Get(ctx, tokKey(t.TenantID)).Result()
		used, _ := strconv.ParseInt(raw, 10, 64)
		if used >= *cfg.TokenLimit {
			return errs.RateLimit("Daily token quota exceeded",
				map[string]any{"used": used, "limit": *cfg.TokenLimit})
		}
	}
	// 2) atomic INCR for request count
	if cfg.RequestLimit != nil {
		n, err := rc.Incr(ctx, reqKey(t.TenantID)).Result()
		if err != nil {
			return err
		}
		if n == 1 {
			_ = rc.Expire(ctx, reqKey(t.TenantID), 48*time.Hour).Err()
		}
		if int(n) > *cfg.RequestLimit {
			return errs.RateLimit("Daily request quota exceeded",
				map[string]any{"used": n, "limit": *cfg.RequestLimit})
		}
	}
	return nil
}

// IncrementTokens is called by usage recording after the upstream reply.
func IncrementTokens(ctx context.Context, tenantID int64, total int64) error {
	if total <= 0 {
		return nil
	}
	rc := redisx.C()
	n, err := rc.IncrBy(ctx, tokKey(tenantID), total).Result()
	if err != nil {
		return err
	}
	if n == total {
		_ = rc.Expire(ctx, tokKey(tenantID), 48*time.Hour).Err()
	}
	return nil
}

// Usage reports today's used/limit for both meters. Used by GET /api/me/quota.
type Usage struct {
	Date     string `json:"date"`
	Requests struct {
		Used  int64  `json:"used"`
		Limit *int   `json:"limit"`
	} `json:"requests"`
	Tokens struct {
		Used  int64  `json:"used"`
		Limit *int64 `json:"limit"`
	} `json:"tokens"`
}

func ReadUsage(ctx context.Context, tenantID int64, plan string) (*Usage, error) {
	cfg, err := loadEffective(ctx, tenantID, plan)
	if err != nil {
		return nil, err
	}
	rc := redisx.C()
	rRaw, _ := rc.Get(ctx, reqKey(tenantID)).Result()
	tRaw, _ := rc.Get(ctx, tokKey(tenantID)).Result()
	rUsed, _ := strconv.ParseInt(rRaw, 10, 64)
	tUsed, _ := strconv.ParseInt(tRaw, 10, 64)
	u := &Usage{Date: todayKeyCN()}
	u.Requests.Used, u.Requests.Limit = rUsed, cfg.RequestLimit
	u.Tokens.Used, u.Tokens.Limit = tUsed, cfg.TokenLimit
	return u, nil
}

// InvalidateCache forces a re-read of the per-tenant override on next Enforce.
func InvalidateCache(ctx context.Context, tenantID int64) error {
	return redisx.C().Del(ctx, cfgKey(tenantID)).Err()
}

func ptrI64(v int64) *int64 { return &v }
func ptrInt(v int) *int     { return &v }
